Source/Core/DolphinQt2/MainWindow.cpp

```diff
@@ -478,3 +493,105 @@ void MainWindow::SetStateSlot(int slot)
   Settings::Instance().SetStateSlot(slot);
   m_state_slot = slot;
 }
+
+void MainWindow::PerformOnlineUpdate(const std::string& region)
+{
+  const int confirm = QMessageBox::question(
+      this, tr("Confirm"), tr("Connect to the Internet and perform an online system update?"));
+  if (confirm != QMessageBox::Yes)
+    return;
+
+  // Do not allow the user to close the dialog. Instead, wait until the update is finished
+  // or cancelled.
+  class UpdateProgressDialog final : public QProgressDialog
+  {
+  public:
+    using QProgressDialog::QProgressDialog;
+
+  protected:
+    void reject() override {}
+  };
+
+  UpdateProgressDialog dialog;
+  dialog.setLabelText(tr("Preparing to update...\nThis can take a while."));
+  dialog.setWindowTitle(tr("Updating"));
+  // QProgressDialog doesn't set its minimum size correctly.
+  dialog.setMinimumSize(360, 150);
+
+  // QProgressDialog doesn't allow us to disable the cancel button when it's pressed,
+  // so we have to pass it our own push button. Note: the dialog takes ownership of it.
+  auto* cancel_button = new QPushButton(tr("&Cancel"), this);
+  dialog.setCancelButton(cancel_button);
+  Common::Flag was_cancelled;
+  disconnect(&dialog, SIGNAL(canceled()), &dialog, SLOT(cancel()));
+  connect(&dialog, &QProgressDialog::canceled, [&] {
+    dialog.setLabelText(tr("Finishing the update...\nThis can take a while."));
+    cancel_button->setEnabled(false);
+    was_cancelled.Set();
+  });
+
+  std::future<WiiUtils::UpdateResult> result =
+      std::async(std::launch::async, &WiiUtils::DoOnlineUpdate,
+                 [&dialog, &was_cancelled](size_t processed, size_t total, u64 title_id) {
+                   Core::QueueHostJob(
+                       [&dialog, &was_cancelled, processed, total, title_id]() {
+                         if (was_cancelled.IsSet())
+                           return;
+
+                         dialog.setRange(0, static_cast<int>(total));
+                         dialog.setValue(static_cast<int>(processed));
+                         dialog.setLabelText(tr("Updating title %1...\nThis can take a while.")
+                                                 .arg(title_id, 16, 16, QLatin1Char('0')));
+                       },
+                       true);
+                   return !was_cancelled.IsSet();
+                 },
+                 region);
+
+  dialog.open();
+  // We manually run the event loop here to avoid the GUI becoming unresponsive.
```

Comment:
Running the event loop manually still seems sketchy to me :/ Would something like this work?

```c++
std::future<WiiUtils::UpdateResult> result = std::async(std::launch::async, [&dialog, &was_cancelled, region] {
  WiiUtils::UpdateResult result_code = WiiUtils::DoOnlineUpdate(
      [&dialog, &was_cancelled](size_t processed, size_t total, u64 title_id) {
        Core::QueueHostJob(
            [&dialog, &was_cancelled, processed, total, title_id]() {
              if (was_cancelled.IsSet())
                return;
              dialog.setRange(0, static_cast<int>(total));
              dialog.setValue(static_cast<int>(processed));
              dialog.setLabelText(tr("Updating title %1...\nThis can take a while.")
                                      .arg(title_id, 16, 16, QLatin1Char('0')));
            },
            true);
        return !was_cancelled.IsSet();
      },
      region);
  dialog.close();
  return result_code;
});
dialog.exec();
```